File: MATCH/177DIV2/DD.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int A[8];
int max;
int ans;

void f(int index, int num) {
    A[index] = num;
    int flag = 0;
    if (index == max) {
        for (int i = 0; i < num; ++ i) {
            int t = A[i];
            if (i == 0 && A[0] == 0) 
                flag = 1;
            else for (int j = 0; j < 8; ++ j) {
                t = A[t];
                if (t == 0) 
                    break;
            }
            if  ( t != 0) {
                flag = 1;
                break;
            }
        }

        if (!flag) {
            ans += 1;
        }

    }
    else {
        for (int i = 0; i < 8; ++ i) 
            f(index + 1, i);
    }
}

int main() {
    for (int i  = 0; i < 8; ++ i) {
        max = i;
        ans = 0;
        f(0, i);
        printf("%d, ", ans + 1);
    }

    return 0;
}

File: contests/177DIV2/DD.py

ans = 0
A = [0]

def f(x, m):
    global A
    A.append(x)
    ff = 0
    if len(A) - 1 == m:
        for i in A:
            t = i
            flag = 0
            for i in xrange(8):
                t = A[t]
                if t == 1:
                    flag = 1
                    break
            if flag == 0:
                ff = 1
                break
        if ff == 0:
            global ans
            ans += 1
    else:
        for i in xrange(m):
            f(i + 1, m)

for i in xrange(8):
    global ans
    ans = 0
    x = i + 1
    for j in xrange(x):
        f(j + 1, x)
    print ans


